definitions/marts/looker_studio_marts/ecf/ls_ecf_correct_cohort_checker.sqlx

```diff
@@ -0,0 +1,237 @@
+
+  Instructions:
+    For ECTs in the "wrong" cohort with declarations:
+      WHERE
+        eligible_for_funding = TRUE
+        AND
+        participant_role = 'ECT'
+        AND 
+        declaration_state IN ('submitted', 'eligible', 'payable', 'paid')
+        AND
+        niot_partnership_chk = 0
+        AND
+        start_date_suggestion_match = 0
+      
+    For Mentors in the "wrong" cohort with declarations:
+      WHERE
+        eligible_for_funding = TRUE
+        AND
+        participant_role = 'Mentor'
+        AND 
+        declaration_state IN ('submitted', 'eligible', 'payable', 'paid')
+        AND
+        niot_partnership_chk = 0
+        AND
+        earliest_dec_suggestion_match = 0
+
+    For ECTs in the "wrong" cohort without declarations:
+      WHERE
+        eligible_for_funding = TRUE
+        AND
+        participant_role = 'ECT'
+        AND 
+        declaration_state NOT IN ('submitted', 'eligible', 'payable', 'paid')
+        AND
+        has_relevant_declarations = FALSE
+        AND
+        niot_partnership_chk = 0
+        AND
+        start_date_suggestion_match = 0
+
+  Dependencies:
+    dataform.ecf_declarations
+*/
+
+-- Gather ECF Declarations
+WITH declarations AS( 
+  SELECT
+     -- Manually assign the user_id for specific participant profile where the profile no longer exists in latest_cpd data and has no superceeded record to call upon.
+     CASE 
+      WHEN dec.participant_profile_id = '8cd0928c-ecad-482f-95cd-6a0ff3e840ad' THEN '79dee4de-a6dd-407c-ab1c-690fcb6d0056'
+      ELSE dec.user_id
+     END AS user_id
+    ,dec.participant_profile_id
+    ,dec.participant_course AS participant_role
+    ,dec.schedule_identifier
+    ,dec.induction_start_date AS induction_start_date
+    ,dec.cohort
+    ,dec.eligible_for_funding
+    ,dec.partnership_id
+    ,dec.partnership_relationship
+    ,dec.lead_provider_name AS induction_lead_provider_name
+    ,dec.cohort_rolled_over_21_24
+    ,dec.declaration_id
+    ,DATE(declaration_date) AS declaration_date
+    ,dec.declaration_type
+    ,state AS declaration_state
+    ,dec.cpd_lead_provider_name AS lead_provider_name
+    ,MIN(CASE WHEN state IN ('paid', 'payable', 'eligible') THEN DATE(declaration_date) ELSE DATE(2050,12,31) END) OVER (PARTITION BY participant_profile_id) AS earliest_declaration_date
+    ,CASE
+      WHEN SUM(relevant_dec_state) OVER (PARTITION BY participant_profile_id) > 0 THEN TRUE 
+      ELSE FALSE 
+     END AS has_relevant_declarations
+  FROM 
+    (SELECT
+      *,
+      CASE WHEN state IN ('eligible','payable','paid','submitted') THEN 1 ELSE 0 END AS relevant_dec_state
+    FROM
+      `dataform__TP_Dev.ecf_declarations`
+    ) dec
+),
+
+-- Add on the grouping of suggested cohorts by induction start date and earliest declaration, Partnership-NIOT check flag, unique participant flag and row number for duplication removal of nulled declarations. 

```

Comment:
Is there a reason we have more than two subqueries in this mart? It appears we could have the first query as a CTE which pulls all required fields and adds any necessary checks and then the final query that uses the fields created in the CTE.

Reply:
1 CTE removed and logic created in it has been passed to final output. The reason it was there was due the query I used in the first instance that mean it was possible to take the script and simply add the where clause for a SQL output. In terms of LS usage the CTE was redundant as the filter/where clause is applied within the dashboard.